src/Workspaces/CSharpTest/EmbeddedLanguages/RegularExpressions/CSharpRegexParserTests.cs

```diff
@@ -0,0 +1,291 @@
+        {
+            CheckInvariants(token.LeadingTrivia, ref position, allChars);
+            CheckCharacters(token.VirtualChars, ref position, allChars);
+        }
+
+        private void CheckInvariants(ImmutableArray<RegexTrivia> leadingTrivia, ref int position, ImmutableArray<VirtualChar> allChars)
+        {
+            foreach (var trivia in leadingTrivia)
+            {
+                CheckInvariants(trivia, ref position, allChars);
+            }
+        }
+
+        private void CheckInvariants(RegexTrivia trivia, ref int position, ImmutableArray<VirtualChar> allChars)
+        {
+            switch (trivia.Kind)
+            {
+            case RegexKind.CommentTrivia:
+            case RegexKind.WhitespaceTrivia:
+                break;
+            default:
+                Assert.False(true, "Incorrect trivia kind");
+                return;
+            }
+
+            CheckCharacters(trivia.VirtualChars, ref position, allChars);
+        }
+
+        private static void CheckCharacters(ImmutableArray<VirtualChar> virtualChars, ref int position, ImmutableArray<VirtualChar> allChars)
+        {
+            for (var i = 0; i < virtualChars.Length; i++)
+            {
+                Assert.Equal(allChars[position + i], virtualChars[i]);
+            }
+
+            position += virtualChars.Length;
+        }
+
+        [Fact]
+        public void TestDeepRecursion()
+        {
+            var (token, tree, chars) =
+                JustParseTree(
+@"@""((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((
+(((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((""", RegexOptions.None, conversionFailureOk: false);
+            Assert.False(token.IsMissing);
+            Assert.False(chars.IsDefaultOrEmpty);
+            Assert.Null(tree);
+        }
+    }
+}
```

Comment:
Maybe I missed it, but do we check round-tripping?

Reply:
I don't.  But that's a good idea.  I will add that to the tests!

Reply:
Oh, nevermind.  I do that test that.  Those are the CheckInvariants and CheckCharacters calls that we make.

Reply:
I'd prefer to not create a ToString (yet) that isn't used anywhere but tests.  We already enforce hte invariants here through the supported API.